update/sync/syncTransactions_test.go

```diff
@@ -341,3 +345,585 @@ func TestSyncPendingTransactionsFor_ReceiveMissingTx(t *testing.T) {
 }
+
+func TestTransactionsSync_RequestTransactionsForPeerMiniBlockShouldWork(t *testing.T) {
+	t.Parallel()
+
+	svi1 := &state.ShardValidatorInfo{
+		PublicKey: []byte("x"),
+	}
+	svi2 := &state.ShardValidatorInfo{
+		PublicKey: []byte("y"),
+	}
+
+	args := createMockArgs()
+	args.DataPools = &dataRetrieverMock.PoolsHolderStub{
+		ValidatorsInfoCalled: func() dataRetriever.ShardedDataCacherNotifier {
+			return &testscommon.ShardedDataStub{
+				ShardDataStoreCalled: func(cacheID string) storage.Cacher {
+					return &testscommon.CacherStub{
+						PeekCalled: func(key []byte) (value interface{}, ok bool) {
+							if bytes.Equal(key, []byte("b")) {
+								return svi2, true
+							}
+							return nil, false
+						},
+					}
+				},
+			}
+		},
+	}
+	transactionsSyncer, _ := NewTransactionsSyncer(args)
+
+	miniBlock := &block.MiniBlock{
+		TxHashes: [][]byte{
+			[]byte("a"),
+			[]byte("b"),
+			[]byte("c"),
+		},
+	}
+	transactionsSyncer.mutPendingTx.Lock()
+	transactionsSyncer.mapValidatorsInfo["a"] = svi1
+	transactionsSyncer.mapTxsToMiniBlocks["b"] = miniBlock
+	numMissingValidatorsInfo := transactionsSyncer.requestTransactionsForPeerMiniBlock(miniBlock)
+
+	assert.Equal(t, 1, numMissingValidatorsInfo)
+	assert.Equal(t, 2, len(transactionsSyncer.mapValidatorsInfo))
+	assert.Equal(t, 2, len(transactionsSyncer.mapTxsToMiniBlocks))
+	assert.Equal(t, svi2, transactionsSyncer.mapValidatorsInfo["b"])
+	assert.Equal(t, miniBlock, transactionsSyncer.mapTxsToMiniBlocks["c"])
+	transactionsSyncer.mutPendingTx.Unlock()
+}
+
+func TestTransactionsSync_ReceivedValidatorInfo(t *testing.T) {
+	t.Parallel()
+
+	txHash := []byte("hash")
+	svi := &state.ShardValidatorInfo{
+		PublicKey: []byte("x"),
+	}
+
+	args := createMockArgs()
+	transactionsSyncer, _ := NewTransactionsSyncer(args)
+
+	// stop sync is true
+	transactionsSyncer.receivedValidatorInfo(txHash, svi)
+	transactionsSyncer.mutPendingTx.Lock()
+	assert.Equal(t, 0, len(transactionsSyncer.mapValidatorsInfo))
+	transactionsSyncer.mutPendingTx.Unlock()
+
+	// txHash does not exist in mapTxsToMiniBlocks
+	transactionsSyncer.stopSync = false
+	transactionsSyncer.receivedValidatorInfo(txHash, svi)
+	transactionsSyncer.mutPendingTx.Lock()
+	assert.Equal(t, 0, len(transactionsSyncer.mapValidatorsInfo))
+	transactionsSyncer.mutPendingTx.Unlock()
+
+	miniBlock := &block.MiniBlock{
+		TxHashes: [][]byte{
+			[]byte("a"),
+			[]byte("b"),
+			[]byte("c"),
+		},
+	}
+	transactionsSyncer.mutPendingTx.Lock()
+	transactionsSyncer.mapTxsToMiniBlocks[string(txHash)] = miniBlock
+	transactionsSyncer.mutPendingTx.Unlock()
+
+	// value received is not of type *state.ShardValidatorInfo
+	transactionsSyncer.receivedValidatorInfo(txHash, nil)
+	transactionsSyncer.mutPendingTx.Lock()
+	assert.Equal(t, 0, len(transactionsSyncer.mapValidatorsInfo))
+	transactionsSyncer.mutPendingTx.Unlock()
+
+	wasReceivedAll := false
+	go func() {
+		select {
+		case <-transactionsSyncer.chReceivedAll:
+			wasReceivedAll = true
```

Comment:
as discussed, this needs to be protected or simply use atomic.Flag implementation

Reply:
Done